### ISSUE TITLE
Remove PyPAM support from Twisted

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3728]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3728|
|Type|defect|
|Created|2009-03-30 23:37:08Z|
|Branch|https://github.com/twisted/twisted/tree/remove-pamauth-3728-2|

The way cred hooks into pam is like so:

  1. launch a thread to talk to the blocking pam api
  1. in the thread, set the euid to root
  1. in the thread, call a blocking api and wait for it to finish
  1. meanwhile, the whole process is running with root privileges
  1. eventually the blocking api finishes, the original euid is restored, the result is returned to the reactor thread

ZOMG!@  NO!

Attachments:

* [pamdemo.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0d6/0d6abd179d351177b0cc63daeec32a8664dc6795/c4dd42f7024d5c17a890e5ca388be46944f3b13c.py) (640 bytes) - added by exarkun on 2009-07-06 18:38:33Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3728 3728
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__remove_pamauth_3728_2 remove-pamauth-3728-2
branch_author__hawkowl__tomprince hawkowl, tomprince
status__closed closed
resolution__fixed fixed
component__core core
keywords__security security
time__1238456228000000 1238456228000000
changetime__1466008476302013 1466008476302013
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to @exarkun|
|-|-|

So, what's the _right_ way to do this, given that the API presented is a blocking one?  Spawn a subprocess and do authentication there?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

A child process is probably the only way to get the correct privilege separation, yes.  I have some other issues with `PluggableAuthenticationModulesChecker` as well, foremost that it is either broken (ie, you cannot authenticate against PAM using it) or has undocumented PAM configuration requirements.  Attached is an example which should do something, but instead always fails with a "conversation error" for me.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

As far as I can tell there is absolutely no documentation for PyPAM.  However, based on the program exarkun attached above, I experimentally determined that the "conversation" callback expects _three_ arguments, not one.  The first is the `PAM.pam` object itself. The second is the "items" dictionary that is currently expected, and the third is always `None` as far as I can tell.

This trivial patch will fix the checker so that it will do _something_, but the aforementioned problems with privilege separation remain.  Still, hopefully this will be useful as some guidance to someone writing a multiprocess version.

```
#!diff
Index: twisted/cred/pamauth.py
===================================================================
--- twisted/cred/pamauth.py	(revision 27083)
+++ twisted/cred/pamauth.py	(working copy)
@@ -13,7 +13,7 @@
 from twisted.internet import threads, defer
 
 def pamAuthenticateThread(service, user, conv):
-    def _conv(items):
+    def _conv(whatever, items, nevermind):
         from twisted.internet import reactor
         try:
             d = conv(items)
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince commented|
|-|-|

(In [43423]) Branching to deprecate-pamauth-3728.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_7"></a>|@tomprince commented|
|-|-|

I think the PAM checker should just be deprecated. The dependency is not easily installable, and is unmaintained.

 1. Remove `PluggableAuthenticationModulesChecker` from the default set of checkers of `twistd conch`. This is an incompatible change, so needs to follow the [procedure](https://twistedmatrix.com/trac/wiki/CompatibilityPolicy#ProcedureforExceptionstothisPolicy) for such changes.
 2. Deprecate `PluggableAuthenticationModuleChecker` and `twisted.cred.pamauth`.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to @tomprince|
|-|-|

Tom, is this ready for review?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

Assigning it to 15.2, so I remember to either finish up this patch or defer it to 15.3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl commented|
|-|-|

(In [44737]) Branching to remove-pamauth-3728-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl removed owner|
|-|-|

Based off the patch from tomprince.

Builders spun, are all green, please review.

Requires 3 committers to sign off as per the compatibility policy as this breaks backwards compat technically. Please see http://twistedmatrix.com/pipermail/twisted-python/2015-May/029478.html for more details.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

+1 from me, as I said on the mailing list.

**Comment by twisted-trac** [comment]
|<img alt="adiroiban1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|adiroiban1 commented|
|-|-|

+1 for removing it and merging this.

-----------

Maybe we should rename this ticket to "Remove support for PAM" and create a new one called "Add (again) support for PAM" with details about what was wrong with old implementation and how the new support should be implemented. 

I think that initial implementation can use CTypes via python-pam module ... as PyPAM is no longer maintained. Later we can look at implementing low level PAM callback as a C module or using cffi

python-pam module by Chris AtLee was forked to support Python3 https://github.com/FirefighterBlu3/python-pam

I am using a fork of the original code, made by Chris AtLee, on all Linux/Unix operating system. It works and I have buildslaves testing the code on Linux/AIX/Solaris/OSX/HPUX https://github.com/chevah/python-pam

---------

Maybe instead of deferToThread we can use something like deferToPoolRootProcess or deferToPoolProcess(uid, gid) which can be implemented based on something similar to deferToAMPProcess from ampoule... this can be a separate ticket.

We can use this root process for replacing twisted.python.util.runAsEffectiveUser

I think that we should also fix/deprecate/remove runAsEffectiveUser as it looks to me that it is vulnerable to the same issue... and so is other code calling os.seteuid/os.setegid like SSHSessionForUnixConchUser and UnixConchUser._runAsuser

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl commented|
|-|-|

Great, I think we have the three voices now. But this needs to remain unmerged for at least a week to give people a chance to see the ML thread.

Untagging 15.2 as it won't make it into this release.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_16"></a>|@hynek set owner to @hawkowl|
|-|-|

LGTM, kill it with fire please.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_17"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44862]) Merge remove-pamauth-3728-2: Remove PAM auth.

Author: tomprince, hawkowl
Reviewers: adiroiban, hynek
Fixes: [#3728](https://github.com/twisted/twisted/issues/3728)
